core/services/llo/telemetry.go

```diff
@@ -0,0 +1,184 @@
+package llo
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/smartcontractkit/libocr/commontypes"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-data-streams/llo"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/llo/evm"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocrcommon"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pipeline/eautils"
+	mercuryutils "github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/mercury/utils"
+	"github.com/smartcontractkit/chainlink/v2/core/services/synchronization/telem"
+)
+
+const adapterLWBAErrorName = "AdapterLWBAError"
+
+type Telemeter interface {
+	EnqueueV3PremiumLegacy(run *pipeline.Run, trrs pipeline.TaskRunResults, streamID uint32, opts llo.DSOpts, val llo.StreamValue, err error)
+}
+
+type TelemeterService interface {
+	Telemeter
+	services.Service
+}
+
+func NewTelemeterService(lggr logger.Logger, monitoringEndpoint commontypes.MonitoringEndpoint) TelemeterService {
+	if monitoringEndpoint == nil {
+		return NullTelemeter
+	}
+	return newTeleter(lggr, monitoringEndpoint)
+}
+
+func newTeleter(lggr logger.Logger, monitoringEndpoint commontypes.MonitoringEndpoint) *telemeter {
+	chTelemetryObservation := make(chan TelemetryObservation, 100)
+	t := &telemeter{
+		chTelemetryObservation: chTelemetryObservation,
+		monitoringEndpoint:     monitoringEndpoint,
+	}
+	t.Service, t.eng = services.Config{
+		Name:  "LLOTelemeterService",
+		Start: t.start,
+	}.NewServiceEngine(lggr)
+
+	return t
+}
+
+type telemeter struct {
+	services.Service
+	eng *services.Engine
+
+	monitoringEndpoint     commontypes.MonitoringEndpoint
+	chTelemetryObservation chan TelemetryObservation
+}
+
+func (t *telemeter) EnqueueV3PremiumLegacy(run *pipeline.Run, trrs pipeline.TaskRunResults, streamID uint32, opts llo.DSOpts, val llo.StreamValue, err error) {
+	var adapterError *eautils.AdapterError
+	var dpInvariantViolationDetected bool
+	if errors.As(err, &adapterError) && adapterError.Name == adapterLWBAErrorName {
+		dpInvariantViolationDetected = true
+	} else if err != nil {
+		// ignore errors
+		return
+	}
+	tObs := TelemetryObservation{run, trrs, streamID, opts, val, dpInvariantViolationDetected}
+	select {
+	case t.chTelemetryObservation <- tObs:
+	default:
+	}
+}
+
+func (t *telemeter) start(_ context.Context) error {
+	t.eng.Go(func(ctx context.Context) {
+		for {
+			select {
+			case tObs := <-t.chTelemetryObservation:
+				t.collectV3PremiumLegacyTelemetry(tObs)
+			case <-ctx.Done():
+				return
+			}
+		}
+	})
+	return nil
+}
+
+func (t *telemeter) collectV3PremiumLegacyTelemetry(d TelemetryObservation) {
+	eaTelemetryValues := ocrcommon.ParseMercuryEATelemetry(t.eng.SugaredLogger, d.trrs, mercuryutils.REPORT_V3)
```

Comment:
Are we going to have to support v4 here? I'm happy to add the support if so.

Reply:
Parallel comp does not support v4 at this time. We will need an additional piece of work in Q4 to add that.